src/Umbraco.Infrastructure/Services/Implement/DistributedJobService.cs

```diff
@@ -0,0 +1,96 @@
+﻿using Microsoft.Extensions.Logging;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.Scoping;
+using Umbraco.Cms.Infrastructure.BackgroundJobs;
+using Umbraco.Cms.Infrastructure.Models;
+using Umbraco.Cms.Infrastructure.Persistence.Repositories;
+
+namespace Umbraco.Cms.Infrastructure.Services.Implement;
+
+/// <inheritdoc />
+public class DistributedJobService : IDistributedJobService
+{
+    private readonly ICoreScopeProvider _coreScopeProvider;
+    private readonly IDistributedJobRepository _distributedJobRepository;
+    private readonly IEnumerable<IDistributedBackgroundJob> _distributedBackgroundJobs;
+    private readonly ILogger<DistributedJobService> _logger;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DistributedJobService"/> class.
+    /// </summary>
+    /// <param name="coreScopeProvider"></param>
+    /// <param name="distributedJobRepository"></param>
+    /// <param name="distributedBackgroundJobs"></param>
+    /// <param name="logger"></param>
+    public DistributedJobService(
+        ICoreScopeProvider coreScopeProvider,
+        IDistributedJobRepository distributedJobRepository,
+        IEnumerable<IDistributedBackgroundJob> distributedBackgroundJobs,
+        ILogger<DistributedJobService> logger)
+    {
+        _coreScopeProvider = coreScopeProvider;
+        _distributedJobRepository = distributedJobRepository;
+        _distributedBackgroundJobs = distributedBackgroundJobs;
+        _logger = logger;
+    }
+
+    /// <inheritdoc />
+    public async Task<IDistributedBackgroundJob?> TryTakeRunnableAsync()
+    {
+        using ICoreScope scope = _coreScopeProvider.CreateCoreScope();
+
+        scope.EagerWriteLock(Constants.Locks.DistributedJobs);
```

Comment:
Just a question, why are we using eager locks here?

Reply:
This is mostly to ensure that we have control over the table while we're reading AND updating, otherwise, we could run into cases where other servers also read the job table at the same time we're doing it, and might end up taking the same jobs (just a thought in my head, have no way of proving it 🙈)